examples/klog_as/klog_as_test.go

```diff
@@ -0,0 +1,74 @@
+// Copyright 2021 The Kubernetes Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package klog_test
+
+import (
+	"flag"
+	"fmt"
+	"os"
+
+	"github.com/go-logr/logr"
+
+	"k8s.io/klog/v2"
+)
+
+type myStruct struct {
+	some, data string
+}
+
+func ExampleAs() {
+	klog.InitFlags(nil)
+	flag.Set("logtostderr", "false")
+	flag.Set("alsologtostderr", "false")
+	flag.Set("skip_headers", "true")
+	klog.SetOutput(os.Stdout)
+
+	item := myStruct{
+		some: "thing",
+		data: "someone",
+	}
+	klog.InfoS("No formatting", "item", item)
+
+	as := klog.As(
```

Comment:
I don't think I understand the intermediate structs.  I thought about something like this in th past but let it slip.  Why is it more complex than:

```
// maybe in logr?

func AsString(fn func() string) fmt.Stringer {
    return asString{fn}
}

func AsStruct(fn func() logr.PseudoStruct {
    return asStruct{fn}
}
```

```
// in caller

log.V(3).Info("the message", "arg", logr.AsText(func() string {
    // complex or expensive or hyper-local logic
})
```

I'm not even sure we need different functions?  Why not:

```
type LogFunc func() interface{}
``

and

```
log.V(3).Info("the message", "arg", logr.LogFunc(func() interface{}) {
    return expensiveProcessing(obj)
}
```

and handle LogFunc as a parallel to Stringer and LogMarshaller

Reply:
> `log.V(3).Info("the message", "arg", logr.AsText(func() string`

This will always be rendered as text, both in text output and JSON. Such a call can be useful to avoid extra `Enabled()` calls. It's not absolutely required.

> `log.V(3).Info("the message", "arg", logr.LogFunc(func() interface{})`

This has two problems:
- only loggers which support `logr.Marshaler` will invoke the function; the klog text format will call `fmt.Sprintf("%v", <proxy object>)`, which then renders as `object(0x<address>)`
- if it gets called, the function cannot decide whether the caller wants a human-readable string (text format) or machine-readable struct (JSON)